LUCENE-9963 Fix issue with FlattenGraphFilter throwing exceptions from holes

glawson0 opened the pull request:
<!--
_(If you are a project committer then you may remove some/all of the following template.)_

Before creating a pull request, please file an issue in the ASF Jira system for Lucene:

* https://issues.apache.org/jira/projects/LUCENE

You will need to create an account in Jira in order to create an issue.

The title of the PR should reference the Jira issue number in the form:

* LUCENE-####: <short description of problem or changes>

LUCENE must be fully capitalized. A short description helps people scanning pull requests for items they can work on.

Properly referencing the issue in the title ensures that Jira is correctly updated with code review comments and commits. -->


# Description

There are various situations were the flatten graph filter may throw an exception or drop tokens due to mishandling of holes in the analysis graph. The primary causes were
1) Output nodes that were hole destinations and valid sources
2) Freeing nodes before they had been fully resolved
3) Putting edges at incorrect output nodes after a hole
# Solution

1)Iterate through all output.InputNodes since dest gaps can exist
2)Freebefore the minimum input node not the first input node(which is usually, but not always, the minimum)
3)Recovery nodes that have end offset less than their preceeding node gaplessly follow their preceeding node
4)Ensure nodes  positionInc >1 are arriving at the end of the graph.

# Tests

same tests as https://github.com/apache/lucene/pull/146
plus tests for shingled tokens and gaps at the start of a token graph.

# Checklist

Please review the following and check all that apply:

- [x] I have reviewed the guidelines for [How to Contribute](https://wiki.apache.org/lucene/HowToContribute) and my code conforms to the standards described there to the best of my ability.
- [x] I have created a Jira issue and added the issue ID to my pull request title.
- [x] I have given Lucene maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [x] I have developed this patch against the `main` branch.
- [x] I have run `./gradlew check`.
- [x] I have added tests for my changes.

Commit:
LUCENE-9963 Add tests for alternate path failures in FlattenGraphFilter

Currently failing tests marked AwaitingFix so they will be ignored.

Commit:
add random test, improve comments

Commit:
LUCENE-9963 Fix issue with FlattenGraphFilter throwing exceptions from holes

4 main ideas as part of this fix
1)Iterate through all output.InputNodes since dest gaps can exist
2)Freebefore the minimum input node not the first input node(which is usually, but not always, the minimum)
3)Recovery nodes that have end offset less than their preceeding node gaplessly follow their preceeding node
4)Ensure nodes  positionInc >1 are arriving at the end of the graph.

glawson0 commented on lucene/analysis/common/src/java/org/apache/lucene/analysis/core/FlattenGraphFilter.java:
```diff
@@ -193,14 +194,25 @@ private boolean releaseBufferedToken() {
           outputFrom++;
-          inputNodes.freeBefore(output.inputNodes.get(0));
+          int freeBefore = Collections.min(output.inputNodes);
+          assert outputNodes.get(outputFrom).inputNodes.stream().filter(n -> freeBefore < n).count()
```
I'm nervous about iterating this list. It's going to scale with how many edges arrive at the same node in the output which _shouldn't_ be large. But that's a big risk for an assert

zhaih replied:
I'm a bit confused about this assertion. Are we trying to assert no future node will be freed mistakenly? If that's the case, shouldn't we assert:
```
assert outputNodes.get(outputFrom).inputNodes.stream().filter(n -> freeBefore > n).count() == 0
    : "FreeBefore " + freeBefore + " will free in use nodes"
```

glawson0 replied:
You're correct. The test as written here just checks if at least one node is ok instead of all nodes, then prints the wrong message.

zhaih commented on lucene/analysis/common/src/java/org/apache/lucene/analysis/core/FlattenGraphFilter.java:
```diff
@@ -240,7 +252,13 @@ private boolean releaseBufferedToken() {
+                        .filter(n -> freeBefore < n)
+                        .count()
+                    > 0
+                : "FreeBefore " + output.inputNodes.get(0) + " will free in use nodes";
```
Same question?

zhaih commented on lucene/analysis/common/src/java/org/apache/lucene/analysis/core/FlattenGraphFilter.java:
```diff
@@ -362,6 +378,40 @@ public boolean incrementToken() throws IOException {
+
+    int maxOutIndex = outputNodes.getMaxPos();
+    OutputNode outSrc = outputNodes.get(maxOutIndex);
+    // There are two types of holes, neighbor holes and consumed holes. A neighbor hole is between
```
Could you add some example explaining these 2 types of holes? 
Seems to me that consumed holes are like 
```
abc: [0,3]
b: [1,2] => c: [2,3]
``` 
right? But I didn't really get what neighbor hole looks like...

zhaih replied:
Thank you, that helps a lot!

zhaih reviewed (commented):
Thank you for making this PR! This is really a complicated filter and I can imagine how hard it is to fix those edge cases.

I wonder whether it is possible to link those edge cases to the specific test cases you added, that should make the review process easier I think?

Commit:
add comments about changed sections. Reference tests in comments

msokolov reviewed (approved):
Somehow the way I was looking at it in github I could only see a few comments and refactoring, but now I see all the new test cases too. Let's clean up the nits in the comments, then I think we can push

zhaih commented on lucene/analysis/common/src/java/org/apache/lucene/analysis/core/FlattenGraphFilter.java:
```diff
@@ -362,6 +394,48 @@ public boolean incrementToken() throws IOException {
+    // of the long token as that's how we'd resolve it if the missing token were present.
+    // neighbor holes should start a new output node and continue as if the hole didn't
+    // exist.
+    // Related tests testAltPathLastStepHoleFollowedByHole, testAltPathFirstStepHole,
```
Thank you for linking the tests here!

zhaih commented on lucene/analysis/common/src/java/org/apache/lucene/analysis/core/FlattenGraphFilter.java:
```diff
@@ -255,6 +260,32 @@ private boolean releaseBufferedToken() {
+    int freeBefore = Collections.min(output.inputNodes);
+    // This will catch a node being freed early if it's input to the next output.
+    // Could a freed early node be input to a later output?
+    assert outputNodes.get(outputFrom).inputNodes.stream().filter(n -> freeBefore < n).count() > 0
+        : "FreeBefore " + output.inputNodes.get(0) + " will free in use nodes";
```
Isn't this still the old assertion that need to be changed?

mikemccand commented:
I love all the added test cases!

Maybe in the new random test we could use `tokenstreamToAutomaton` and then confirm no dead states in that, using existing APIs?

mikemccand commented:
Maybe another way to improve the checking for correctness in the randomized test (or maybe in a new randomized test) would be to randomly generate a set of strings from a limited alphabet, create the minimal automaton matching only those strings (we have a nice API to do that, efficiently, already), call flatten, and the confirm that the resulting output graph still accepts all the original strings?

I.e., flatten should only ever "generalize" -- accepting strings that the original machine did not -- and never "remove" previously accepted strings?

But I think one missing part for such a test would be an "Automaton to TokenStream" converter, i.e. a "serializer" from (acyclic) Automaton to TokenStream.  I think such a thing would not be too difficult to build, basically just topo sort the input graph (and throw exception if it has cycles), then emit the transitions as tokens.  The `posInc` attribute is guaranteed to never go negative because of the topo sort.  This would (separately) be a nice utility API to convert between these two things that are really nearly the same ;)

mikemccand commented:
Thank you for the ascii-art explanation of neighbor versus consumed holes!

Do we also handle the neighbor-hole case when there is more than one long path?

E.g. if the document has `the wizard of oz`, and we overlay `the_wizard` (pos len 2) and also `the_wizard_of_oz` (pos len 4) shingles on top of that, then run through `StopFilter` which deletes `the` (hmm and also `of`, but ignore that consumed hole for now), then we have a neighbor hole with three paths (the original tokens plus the two shingles).  Do we handle that / have a test case for it?

mikemccand reviewed (commented):
I left smallish comments so far -- still trying to remember how this code works before giving a better review!

glawson0 commented:
>Maybe another way to improve the checking for correctness in the randomized test (or maybe in a new randomized test) would be to randomly generate a set of strings from a limited alphabet, create the minimal automaton matching only those strings (we have a nice API to do that, efficiently, already), call flatten, and the confirm that the resulting output graph still accepts all the original strings?

Are you referring to the Automoton.Builder API here?

> But I think one missing part for such a test would be an "Automaton to TokenStream" converter, i.e. a "serializer" from (acyclic) Automaton to TokenStream. I think such a thing would not be too difficult to build, basically just topo sort the input graph (and throw exception if it has cycles), then emit the transitions as tokens. The `posInc` attribute is guaranteed to never go negative because of the topo sort. This would (separately) be a nice utility API to convert between these two things that are really nearly the same ;)

I'm having trouble with the transition from Automoton back to TokenStream. `posInc` is easy to calculate because after a toposort it's either 1 or 1+ number of holes since last non-hole. `poslen` is giving me some trouble because it needs to know how many positions were in alternate paths. I should be able to look at each path from a node that has multiple edges, follow those paths until they converge counting token separators along the way, then take the max of those as the length to be covered. However I'm not sure how to distribute that length across paths.
For example
```
 --->O--->O--->O---
|                  |
|                  V
(Start)           (End)
|                  ^
----->O------>O-----
```
There are 3 positions on the longest path so the total positions to be covered is 3, but on the short 2 node path I don't think it's possible to tell what should be `poslen`=1 and which should be `poslen`=2. For creating a valid token stream that can match the same as the automaton can accept it might not matter. However to reverse `tokenstreamToAutomaton` we would need to get those `poslen`s right.
We would also need to set Offsets. Offsets are completely lost in the conversion from `TokenStream` to `Automaton`.  We should be able to generate offsets that are valid for a `TokenStream` but may not reflect what the text originally looked like. For creating a general `AutomotonToTokenStream` I think this would be a problem, as `tokenstreamToAutomaton` -> `AutomotonToTokenStream` would end in a different `TokenStream` then one started with.

For the generalize test I might have to create an `Automaton` from a `TokenStream` with and one without `FlattenGraphFilter`, then create the acceptable paths from the one without `FGF` and check if the `Automaton` with `FGF` can accept all of them. It's less elegant but would still confirm that the flattened graph was a `generalization`.

Commit:
adding generalization tests, change recover from hole to pick output nodes based off posinc, don't free from hole sources, and make gap cross check based on minToNode.

glawson0 commented:
While writing the generalization test I've realized there are some strange interactions with it and holes. For example the `FlattenGraphFilter` will squash continuous sequences of holes down to one. Also if all the edges in a path get deleted then `FlattenGraphFilter` will squash the remaining edges down to remove the deleted path.

I think the first aspect of squashing down holes may be a bug. I noticed the behavior earlier and added a test for it, but looking at the code it may come from hole recovery using `outputNodes.getMaxPos() + 1` instead of `outputNodes.getMaxPos() + positionIncrement`

mikemccand commented:
> For the generalize test I might have to create an `Automaton` from a `TokenStream` with and one without `FlattenGraphFilter`, then create the acceptable paths from the one without `FGF` and check if the `Automaton` with `FGF` can accept all of them. It's less elegant but would still confirm that the flattened graph was a `generalization`.

Ooh +1 this is a good test too!  But you would need some way to randomly creating "interesting" TokenStreams.  Hmm do we have a `RandomGraphTokenStream` yet?

mikemccand commented:
> However I'm not sure how to distribute that length across paths

I think you are making it harder than it needs to be!

First, topo sort the `Automaton`, which results in integer numbers assigned to each node (different integers than the original states, in general).

Then you will output each transition from the `Automaton` as a token in the `TokenStream`, one by one.  Start with the state numbered `0` (by the toposort).  Output all transitions leaving it.  For `posLength` of each of those, just look at the `to` topo-sort state and subtract those two integers.  E.g. if topo-sort state `0` has a transition to topo-sort state `7` then that token's `posLength` is `7`.

Give a non-zero `posInc` each time you jump to the next top-sort state, and pretend you jumped from state=-1 for the first state=0.

Note that this algorithm requires a finite Automaton (we should call `Operations.isFinite` to ensure, and throw an exception if it is infinite).  Also, it cannot handle fully disconnected graphs ("isolated subgraphs").  All nodes must be reachable from any other node.  If the randomized test simple creates the DFA accepting a set of randomly generated strings both of these properties will hold.

glawson0 commented on lucene/analysis/common/src/java/org/apache/lucene/analysis/core/FlattenGraphFilter.java:
```diff
@@ -284,27 +316,27 @@ public boolean incrementToken() throws IOException {
-          // case for us because we normally rely on the full dependencies expressed
-          // by the arcs to assign outgoing node IDs.  It would be better if tokens
-          // were never dropped but instead just marked deleted with a new
-          // TermDeletedAttribute (boolean valued) ... but until that future, we have
```
Working on this change has made me realize how valuable not removing tokens would be.

glawson0 commented:
Thanks for the explanation. I think I understand how to do the `Automaton` to `TokenStream` now.

glawson0 commented:
Changed 3 parts of flatten graph filter
1) don't call freeBefore when we finish a hole src node. Hole src nodes can mean something hasn't been accounted for correctly so calling freeBefore may free nodes that still have info we need.
2) Change how RecoverFromHole picks the outSrc node. Instead of always always adding to the end of the OutputNodes list it scales down its position increment to match the output graph and uses that to pick the OutputNode. This doesn't require the distinction between neighbor and consumed holes.
3) Change the condition for fixing gaps after an alternate path hole. Now it checks if the shortest edge from the position the current edge is incrementing from comes to the current start node. if it doesn't then we've crossed a gap.


Added a few more tests `testHeavilyConnectedGraphWithGap` and `testShingleFromGap` I believe cover the cases @mikemccand mentioned earlier. I also added the `Automaton` -> `TokenStream` -> Flattened`TokenStream` -> Automaton no paths lost test as well as a validation test on the existing random graph test. The validation gets weird with handling holes, but I think that's necessary as a hole in an accept sequence could be anything and hole sizes can change after flattening.

glawson0 commented:
For additional testing I've built a few indexes and played some queries against them. I didn't find any changes to latency or memory usage. Depending on the data set used there was impact on results. The English data sets didn't show any impact. Using it with our Japanese analyzer had a larger impact. In one index 10% of queries were impacted by the change with those queries matching more documents. Previously those documents had lost tokens from `FlattenGraphFilter`. The primary cause of these dropped tokens was graphs with gaps in them. The `JapaneseTokenizer` would create graphs that `FilteringTokenFilter`s would remove tokens from and then `FlattenGraphFilter` would incorrectly flatten them.

Commit:
Merge branch 'main' into gap_fix

Commit:
LUCENE-9963: Flatten graph filter has errors when there are holes at beginning or end of alternate paths

    6 main ideas for this fix
    1)Iterate through all output.InputNodes since dest gaps can exist.
    2)freeBefore the minimum input node instead of the first input node(which was usually, but not always, the minimum).
    3)Don't freeBefore from a hole source node. Book keeping may not be correct and could result in an early free.
    4)When adding an output node after hole recovery, calculate its new position increment instead of adding it to the end of the output graph.
    5)Nodes after holes that have edges to their source will do the output re-mapping that the deleted node would have done.
    6)If a disconnected input node swaps order with another node in the output, then map them to the same output node.

glawson0 commented:
Simplified `AutomatonToTokenStream` and added license comment.

I beasted the random tests a few hundred times and found 2 issues.
1)When completely separated edges get connected it can introduce gaps and cause book keeping errors.
2)The `checkGeneralization` test can miss paths where a hole in the unflattened graph would cross multiple nodes.

The logic to navigate around issue 2 slows down `checkGeneralization` to the point I don't think it is useful so I commented that portion out. The test that calls it still checks if asserts are getting hit, and `testPathsNotLost` is checking if strings to make an automaton are accepted after the automaton is flattened.

glawson0 commented:
Failure is from random testing. It looks like I'm overthinking how to handle calculating the new output node in `recoverHole`. Since multiple graphs could look the same after deletions, but would have flattened differently without the deletes, calculating future nodes in the middle of flattening isn't going to work. Instead the code will have to place at the end of the output graph adjusting for known holes.

Commit:
LUCENE-9963: Flatten graph filter has errors when there are holes at beginning or end of alternate paths

    4 main ideas for this fix
    1)Iterate through all output.InputNodes since dest gaps can exist.
    2)freeBefore the minimum input node instead of the first input node(which was usually, but not always, the minimum).
    3)When adding an output node after hole recovery add it to the end of the graph or 1 pos after incoming path.
    4)Nodes after holes that have edges to their source will do the output re-mapping that the deleted node would have done.

glawson0 commented:
We have been using this change internally for a few weeks now. We no longer encounter the `ArrayIndexOutOfBounds` exceptions that we were previously experiencing. Depending on the dataset/analyzer combination we have seen up to a 1% increase in the average number of tokens per field. This comes from the tokens that had previously been dropped now being correctly indexed.

mikemccand commented:
> We have been using this change internally for a few weeks now. We no longer encounter the ArrayIndexOutOfBounds exceptions that we were previously experiencing. Depending on the dataset/analyzer combination we have seen up to a 1% increase in the average number of tokens per field. This comes from the tokens that had previously been dropped now being correctly indexed.

Thanks for the update @glawson0!  This is great news.

mikemccand commented on lucene/core/src/java/org/apache/lucene/analysis/AutomatonToTokenStream.java:
```diff
@@ -0,0 +1,197 @@
+import org.apache.lucene.util.automaton.Transition;
+
+/** Converts an Automaton into a TokenStream. */
+public class AutomatonToTokenStream {
```
Whoa, awesome!  This will be a really helpful infrastructure for future testing!  And of course now I really feel compelled to take a Lev1("lucene") and run it through here and watch what the resulting tokens are!

mikemccand commented on lucene/core/src/java/org/apache/lucene/analysis/AutomatonToTokenStream.java:
```diff
@@ -0,0 +1,197 @@
+    LinkedList<RemapNode> noIncomingEdges = new LinkedList<>();
+    Map<Integer, Integer> idToPos = new HashMap<>();
+    noIncomingEdges.addLast(new RemapNode(0, 0));
+    while (noIncomingEdges.isEmpty() == false) {
```
Whoa, is this toposort?  What algorithm :)  Maybe add a comment that that is what this code is doing, with which algo?

mikemccand commented on lucene/core/src/java/org/apache/lucene/analysis/AutomatonToTokenStream.java:
```diff
@@ -0,0 +1,197 @@
+      } else {
+        positionNodes.get(currState.pos).add(currState.id);
+      }
+      idToPos.put(currState.id, currState.pos);
```
It looks like we "eagerly" place all nodes onto the same position if the topo-sort allows it (i.e. those nodes are at the same "layer" after topo-sort)?  But doesn't that then over-generalize, causing the resulting token graph to accept more strings than the automaton originally did?

mikemccand commented on lucene/core/src/java/org/apache/lucene/analysis/AutomatonToTokenStream.java:
```diff
@@ -0,0 +1,197 @@
+            edges.add(new EdgeToken(destLayer, val));
+            // If there's an intermediate accept state, add an edge to the terminal state.
+            if (automaton.isAccept(t.dest) && destLayer != positionNodes.size() - 1) {
+              edges.add(new EdgeToken(positionNodes.size() - 1, val));
```
Ahh, the token graph has only one "accept" state, the highest end position, so we have to move these accept transitions "out" to that end pos.  Sneaky!

mikemccand commented on lucene/core/src/test/org/apache/lucene/analysis/TestAutomatonToTokenStream.java:
```diff
@@ -0,0 +1,136 @@
+
+    Automaton flatPathAutomaton = DaciukMihovAutomatonBuilder.build(acceptStrings);
+    TokenStream ts = AutomatonToTokenStream.toTokenStream(flatPathAutomaton);
+    assertTokenStreamContents(
```
So cool!!

mikemccand commented on lucene/analysis/common/src/test/org/apache/lucene/analysis/core/TestFlattenGraphFilter.java:
```diff
@@ -544,8 +732,288 @@ protected TokenStreamComponents createComponents(String fieldName) {
+   * @param notFlattened not flattened TokenStream
+   * @throws IOException on error creating Automata
+   */
+  /* private void checkGeneralization(TokenStream flattened, TokenStream notFlattened)
```
Hmm why did you need to comment out all these methods?

mikemccand commented on lucene/analysis/common/src/java/org/apache/lucene/analysis/core/FlattenGraphFilter.java:
```diff
@@ -362,6 +415,55 @@ public boolean incrementToken() throws IOException {
+        outIndex = inputNodes.get(offsetSrc.minToNode).outputNode + 1;
+      } else {
+        // no information about how the current node was previously connected.
+        // Connect it to the end.
```
This connects the orphan'd token to the "end" as far as we've consumed/produced so far (`outputNodes.getMaxPos()`) right?  This seems OK (not sure how else we could recover) but I wonder in practice what incoming hole configuration tickles this?

mikemccand reviewed (approved):
I left some minor comments.

I REALLY LOVE the new `AutomatonToTokenStream`!  This makes me want to go convert some of our more exotic automata to token stream, like Lev1("lucene")!

zhaih reviewed (approved):
Thanks Geoff, yesterday I've also ran some internal benchmark tools as well as manual document inspection to confirm that this change is doing the right thing.

Thank you for making such a complicated change!

mikemccand commented:
+1, thanks @glawson0 -- I'll try to merge this soon.

I think this is also safe to backport?  No API breaks, and it is "just" fixing bugs in this hairy `TokenFilter`?

glawson0 commented:
Thanks for the merge! You are correct, there are no API changes and it should be safe to backport.
